tmk_core/protocol/chibios/usb_main.c

```diff
@@ -534,40 +530,35 @@ static void usb_event_cb(USBDriver *usbp, usbevent_t event) {
                 }
                 qmkusbConfigureHookI(&drivers.array[i].driver);
             }
-            osalSysUnlockFromISR();
             if (last_suspend_state) {
                 usb_event_queue_enqueue(USB_EVENT_WAKEUP);
             }
```

Comment:
One remaining problem here is that `last_suspend_state` is set in `usb_event_queue_task()` (and without any locking too).  If `usb_event_queue_task()` does not get called in time (which may happen if someone sticks a large delay into some code), we may potentially get here with a stale value for `last_suspend_state`, and therefore lose the wakeup event in case the wakeup goes through the full reconfiguration.  Although getting that case after the other changes in the PR would probably be harder than before (with the older code you could get this broken state for any resume that ends up reconfiguring devices, because `USB_EVENT_SUSPEND` was not processed in time).

The options here are:
1. Move all handling of `last_suspend_state` into `usb_event_cb()` (set it on `USB_EVENT_SUSPEND` and `USB_EVENT_WAKEUP`, and also after generating the `USB_EVENT_WAKEUP` event in the `USB_EVENT_CONFIGURED` path).
2. Move all handling of `last_suspend_state` into `usb_event_queue_task()` (if `USB_EVENT_CONFIGURED` comes after `USB_EVENT_SUSPEND`, but without `USB_EVENT_WAKEUP`, perform wakeup processing too).

Reply:
Thanks for spotting! I did the first approach and moved it into the callback and hopefully didn't miss anything.